### ISSUE TITLE
Cache bug: when repeatedly called a query with the same variables, wrong cache entry returned

### ISSUE THREAD

**Issue by adambisek** [apollographql/react-apollo#1853]
**Dependencies versions:**
"graphql": "^0.13.0",
"graphql-tag": "^2.7.3",   
"react": "^16.2.0",
"react-apollo": "^2.0.4",
"apollo-link": "^1.2.1",

**HoC:**
```
export const withQuery = graphql(query, {
  options: ({ formValues }) => ({requestedValues))),
    variables: {
      rooms: formatRoomsForApollo(formValues.rooms),
    },
    // fetchPolicy: 'network-only', // if I enable network-only, bug is not noticeable anymore
  }),
  props: ({ ownProps, data }) => ({
    ...ownProps,
    termForm: get('termForm')(data) || {},
  }),
})
```

**Query:**
```
query TermFormQuery (
  $rooms: [RoomInput],
) {
  termForm(request: {
    rooms: $rooms,
  }, 
  ) {
    rooms {
      id,
      adults,
      children {
        age
      },
      board,
      room,
      price,
      priceId
    },
  }
}
```

_Note that "board" differs through these three requests._

**First request (used network):**
variables: `rooms:[{id: 1, adults: 2, children: [], board: "HB", room: "APT", price: null}]`
-> result is OK: `[0: {id: 1, adults: 2, children: Array(0), board: "HB", room: "APT", …}]`

**Second request (used network too, because variables changed):**
variables: `rooms:[{id: 1, adults: 2, children: [], board: "BB", room: "APT", price: null}]`
-> result is OK: `[0: {id: 1, adults: 2, children: Array(0), board: "BB", room: "APT", …}]`

**Third request - same as first (with default fetchPolicy may use cache):**
variables: `rooms:[{id: 1, adults: 2, children: [], board: "HB", room: "APT", price: null}]`
-> returns result from cache, but for the 2nd variables, not for the 1st variables
-> result is BAD: `[0: {id: 1, adults: 2, children: Array(0), board: "BB", room: "APT", …}]`

**Intended outcome:**

Third request variables should return same result like in the first query (because third query has same variables like the first one.

**Actual outcome:**

Third request variables returns same result like in the second query, even when variables across second and third query differs. If I disable cache with fetchPolicy, bug dissapears.

**Comment by drFabio** [comment]
I have an error like that but with react-apollo 2.1.1 can you please confirm the exact installed version?

**Comment by adambisek** [comment]
Yes, I can. Version "2.1.1"

**Comment by drFabio** [comment]
I have a related bug in which if I send
```
data(id:1) {
  foo
  bar
}
```
and in otherr query
```
data(id:1) {
  bar
}
```
the first query would lose the foo value if the second query get's resolved after it.

What is solving for me is locking on 2.0.4 can you check if that works for you as well? It seems 2.1.0 might've broken something on the cache

**Comment by harmon25** [comment]
I am having similar trouble when using HOC with new versions. #1833
Reverting to `2.1.0-beta-2` seems to work as intended...

**Comment by adambisek** [comment]
Thanks for info guys. Still hope this is going to be fixed in current stable.

**Comment by rafaponieman** [comment]
I'm experiencing the same issue. I've already refactored and developed on the new Query component, and realized after the fact that it wasn't working. For me, it's not working with 2.1.0-beta-2 or 3 either. I had to resort to fetchPolicy 'network-only'. It's quite important to get the cache back, and now I can't refactor back to 2.0.x because I've alredy done a big implementation on the Query component. Thanks a lot!!
